### PR TITLE
watch Istio resources changes in istio-operator

### PR DESCRIPTION
**Pull request by irisdingbj**
1. Reuse reconciler if  ICP has no change
2. watch delete event for Istio resources and trigger reconcile against ICP
3. Do not update/prune resources if ICP has no change

### CODE REVIEW

File: pkg/controller/istiocontrolplane/istiocontrolplane_controller.go [hunk 1]
```diff
@@ -165,3 +192,101 @@ func (r *ReconcileIstioControlPlane) Reconcile(request reconcile.Request) (recon
...
+		&rbacv1.RoleBinding{TypeMeta: metav1.TypeMeta{Kind: "ClusterRoleBinding", APIVersion: "v1"}},
+		&rbacv1.ClusterRole{TypeMeta: metav1.TypeMeta{Kind: "ClusterRole", APIVersion: "v1"}},
+		&rbacv1.ClusterRoleBinding{TypeMeta: metav1.TypeMeta{Kind: "ClusterRoleBinding", APIVersion: "v1"}},
+		&corev1.ConfigMap{TypeMeta: metav1.TypeMeta{Kind: "ConfigMap", APIVersion: "v1"}},
+		&corev1.Service{TypeMeta: metav1.TypeMeta{Kind: "Service", APIVersion: "v1"}},
+		&appsv1.Deployment{TypeMeta: metav1.TypeMeta{Kind: "Deployment", APIVersion: "v1"}},
+		&appsv1.DaemonSet{TypeMeta: metav1.TypeMeta{Kind: "DaemonSet", APIVersion: "v1"}},
+		&autoscalingv2beta1.HorizontalPodAutoscaler{TypeMeta: metav1.TypeMeta{Kind: "HorizontalPodAutoscaler", APIVersion: "v2beta1"}},
+		&admissionregistrationv1beta1.MutatingWebhookConfiguration{TypeMeta: metav1.TypeMeta{Kind: "MutatingWebhookConfiguration", APIVersion: "v1beta1"}},
+		&v1beta1.Deployment{TypeMeta: metav1.TypeMeta{Kind: "Deployment", APIVersion: "v1beta1"}},
```
Comment (morvencao) [inline pkg/controller/istiocontrolplane/istiocontrolplane_controller.go]: Is it possible to reuse the resources defined at https://github.com/istio/operator/blob/master/pkg/controller/istiocontrolplane/pruningdetails.go#L44-L103 ?

Comment (irisdingbj) [reply pkg/controller/istiocontrolplane/istiocontrolplane_controller.go]: Fixed

**Review by ostromart — APPROVED** [review]

**Comment by irisdingbj** [comment]
e2e failures will be fixed in https://github.com/istio/istio/pull/19509

**Comment by irisdingbj** [comment]
https://github.com/istio/istio/pull/19538

**Pull request merged**